lib/Dialect/FIRRTL/Ops.cpp

```diff
@@ -699,6 +700,93 @@ FIRRTLType MemOp::getDataTypeOrNull() {
 // Statements
 //===----------------------------------------------------------------------===//
 
+static bool areConnectEquivalent(FIRRTLType destType, FIRRTLType srcType);
+
+/// Helper to implement the equivalence logic for a pair of bundle elements in a
+/// connect.
+static bool areElementsConnectEquivalent(
+    std::tuple<BundleType::BundleElement, BundleType::BundleElement> tuple) {
+  BundleType::BundleElement destElement = std::get<0>(tuple);
+  Identifier destElementName = std::get<0>(destElement);
+  FIRRTLType destElementType = std::get<1>(destElement);
+  bool destIsFlip = destElementType.isa<FlipType>();
+
+  BundleType::BundleElement srcElement = std::get<1>(tuple);
+  Identifier srcElementName = std::get<0>(srcElement);
+  FIRRTLType srcElementType = std::get<1>(srcElement);
+  bool srcIsFlip = srcElementType.isa<FlipType>();
+
+  if (destElementName != srcElementName)
+    return false;
+
+  if (destIsFlip != srcIsFlip)
+    return false;
+
+  // If either destination or source is flipped, strip the outer flip for
+  // recursively equivalance checking.
+  if (destIsFlip)
+    destElementType = destElementType.cast<FlipType>().getElementType();
+
+  if (srcIsFlip)
+    srcElementType = srcElementType.cast<FlipType>().getElementType();
+
+  return areConnectEquivalent(destElementType, srcElementType);
+}
+
+/// Returns whether the two types are equivalent for the purpose of
+/// connections. See Section 4.5 in the FIRRTL spec:
+/// https://github.com/freechipsproject/firrtl/blob/master/spec/spec.pdf.
+static bool areConnectEquivalent(FIRRTLType destType, FIRRTLType srcType) {
+  // Analog types cannot be connected and must be attached.
+  if (destType.isa<AnalogType>() || srcType.isa<AnalogType>())
+    return false;
+
+  // Reset types can be driven by UInt<1>, AsyncReset, or Reset types.
+  if (destType.getPassiveType().isa<ResetType>())
+    return srcType.isResetType();
+
+  // Reset types can drive UInt<1>, AsyncReset, or Reset types.
+  if (srcType.isa<ResetType>())
+    return destType.getPassiveType().isResetType();
+
+  // Vector types can be connected if they have the same size and element type.
+  auto destVectorType = destType.dyn_cast<FVectorType>();
+  auto srcVectorType = srcType.dyn_cast<FVectorType>();
+  if (destVectorType && srcVectorType)
+    return destVectorType.getNumElements() == srcVectorType.getNumElements() &&
+           areConnectEquivalent(destVectorType.getElementType(),
+                                srcVectorType.getElementType());
+
+  // Bundle types can be connected if they have the same size, element names,
+  // and element types.
+  auto destBundleType = destType.dyn_cast<BundleType>();
+  auto srcBundleType = srcType.dyn_cast<BundleType>();
+  if (destBundleType && srcBundleType)
+    return destBundleType.getNumElements() == srcBundleType.getNumElements() &&
+           llvm::all_of(llvm::zip(destBundleType.getElements(),
```

Comment:
Please use a for loop instead of zip and all_of here.  The metaprogramming isn't necessary and forces tuples into the type system unnecessarily.

C++ is not a beautiful functional language :-)

Reply:
:) understood... and, i already fixed a bug related to getting the wrong tuple element...